### ISSUE TITLE
Using a 4D attention mask in MultiheadAttention forward

### ISSUE THREAD

**Issue by davidbuterez** [pytorch/pytorch#116472]
### 🚀 The feature, motivation and pitch

The `torch.nn.MultiheadAttention` forward method requires an attention mask of shape `(N * num_heads, L, S)` when the goal is to have a different mask for each item in the batch. I might be missing something but for this shape, it is not possible to use `expand` with the number of heads and then `view`, which would not use any extra memory. Currently, I am using `repeat` which is prohibitively wasteful in some scenarios.

For `torch.nn.functional.scaled_dot_product_attention`, it is possible to use a 4D attention mask based on `expand`, if I remember correctly. Is there a way to use `MultiheadAttention` and an attention mask without allocating any additional memory? If not, could this be considered?

I think @drisspg would be helpful here. I noticed some issues/commits on this topic but I am not up-to-date with the latest developments and subtleties. Thanks!

### Alternatives

The `scaled_dot_product_attention` function and the equivalent `xformers` function seem to support 4D masks. It would be great to have this level of control and memory saving for `MultiheadAttention`.

**Comment by drisspg** [comment]
Out of curiosity is there a reason you prefer to use MHA vs SDPA directly + linears for the in and out projection?

**Comment by davidbuterez** [comment]
@drisspg Mainly because I have observed performance differences between the `MultiheadAttention` module and my code relying on SDPA (most likely I'm doing something wrong, but I'm not convinced as sometimes my SDPA-based implementation is better according to the metrics that I'm tracking). `MultiheadAttention` is also a bit more customizable (e.g. `add_zero_attn`). In any case, being able to significantly reduce memory consumption when using `MultiheadAttention` would surely be a good thing to have?
